### ISSUE TITLE
Linux/PPC Hercules fails with bad RNDGETENTCNT ioctl to /dev/urandom

### ISSUE THREAD

**Issue by friedkiwi** [SDL-Hercules-390/hyperion#132]
When trying to run Hercules on Linux/PPC, hercules hangs with the following output:
```
HHC02322W This build of Hercules has only basic TCP keepalive support
HHC02320W Not all TCP keepalive settings honored
HHC00111I Thread CPU Time IS available (_POSIX_THREAD_CPUTIME=0)
HHC00100I Thread id f76ff440, prio -1, name 'timer_thread' started
HHC00100I Thread id f7936440, prio -1, name 'Processor CP00' started
HHC00811I Processor CP00: architecture mode z/Arch
HHC01494E Crypto: 'ioctl()' failed: Inappropriate ioctl for device
```
Hercules completely hangs at this point - it does not shutdown when you kill it and it needs a `kill -9`.

I've traced the error, and it seems to be that in crypto.c an IOCTL (`RNDGETENTCNT`) is issued to `/dev/urandom` to retrieve the entropy count. The problem is that the value for this constant is hardcoded into hercules and it does not match the value of Linux on PPC (0x40045200).

Ideally, the constant should be sourced from `<linux/crypto.h>`. This issue has been patched in my copy of the repository (see: https://github.com/friedkiwi/hyperion/blob/master/crypto.h). It is safe to swap out the definition of the constant with the include file.

/y

**Comment by Fish-Git** [comment]
> Ideally, the constant should be sourced from <linux/crypto.h>.  `[...]` It is safe to swap out the definition of the constant with the include file.

I'm not so sure, Yvan. I need to do some more research. But I _do_ very much appreciate the suggestion! It sounds like a very good one and one that, ultimately, might turn out to be the correct one.

My only concern is the actual include itself: `<linux/crypto.h>`.

The reason for my concern is that, previously, I had a clone of the Linux 2.2.6 kernel source code that I would use for Hercules research purposes (I have just now deleted it and am in the process of re-cloning the latest and greatest version from [https://github.com/torvalds/linux.git](https://github.com/torvalds/linux.git)), and in my older (now non-existent) Linux 2.2.6 kernel source code tree, the `RNDGETENTCNT` iotctl code is (was) actually defined in the `<linux/random.h>` header (I think), and _not_ the `<linux/crypto.h>` header.

I'll grep the latest Linux source code again (once my git clone completes; it's taking a long time to finish) to confirm whether or not `<linux/crypto.h>` is indeed the correct header to #include or not. If it is, then I will make the change you suggested. If not, then we will, together, have to determine what the _correct_ header to #include actually should be.

But regardless of which header turns out to be correct one, I want to thank you sincerely for reporting this! It's great feedback and I thank you for it!
___
(I wish there was some system API that could be called to tell you what the actual ioctl code actually was, like `sysconf()` or something. I don't suppose there is such a thing, is there? Note: I am _not_ -- repeat _**not**_ a Linux person! I'm a Windoze weenie, so I rely heavily on people such as yourself for proper guidance regarding Linux/Unix issues, so please bear with me while I ask stupid questions in my effort to try and understand things to make sure we're doing the right thing. Thanks.)

**Comment by Fish-Git** [comment]
Addendum: I also want to make clear that the issue regarding which header should be #included is only a _secondary_ issue to be resolved, but is not the _primary_ issue that needs to be resolved.

The _primary_ issue that I need to resolve is to fix my lazy programming in crypto.c that is causing the hang in the first place! _THAT_ should not happen!

I should have coded it more intelligently to automatically "fall back" to using some _other_ method (such as using the `srand` and `rand` APIs) when the primary/preferred method fails for whatever reason.

I will try fix my stupidness as soon as I can and get back to you.

Thanks again for reporting it.

_(grumble-i-hate-it-when-i-do-stupid-shit-grumble...)_

**Comment by Fish-Git** [comment]
> I'll grep the latest Linux source code again ... to confirm whether or not <linux/crypto.h> is indeed the correct header to #include or not.

**Update:** I just downloaded the (current?) Linux source tree from [https://github.com/torvalds/linux.git](https://github.com/torvalds/linux.git) and the only place that `RNDGETENTCNT` is defined is in include/uapi/linux/**random.h**.

So it appears <linux/crypto.h> is the _wrong_ header to #include. It would seem the _correct_ header to be #included (should we even decide to "fix" it that way!) should _actually_ be **random.h** (<linux/random.h>??)

> If not, then we will, together, have to determine what the correct header to #include actually should be.

I'm thinking instead of #including some header to instead simply add some logic to check for `ENOTTY == errno` and if so, to try the alternate value (and then fail if neither works of course). Does that sound reasonable to you @friedkiwi?

I will of course also fix the failure hang to by adding code to fall back to a default srand/rand api technique (along with a warning message of course).

Do both of these seem reasonable to you?

**Comment by Fish-Git** [comment]
@friedkiwi,

> I'm thinking instead of #including some header to instead simply add some logic to check for ENOTTY == errno and if so, to try the alternate value (and then fail if neither works of course). Does that sound reasonable to you @friedkiwi?
> 
> I will of course also fix the failure hang to by adding code to fall back to a default srand/rand api technique (along with a warning message of course).
> 
> Do both of these seem reasonable to you?

The below patch implements both of the above:

* [#crypt_hang.patch.txt](https://github.com/SDL-Hercules-390/hyperion/files/2309572/crypt_hang.patch.txt)

I plan to commit it within the next day or two if I don't hear back from you. Please give it a try and let me know whether or not it fixes your hang. Thanks!

**Comment by Fish-Git** [comment]
> Did you see my email/PR?

I received it yes, but I have not had to time to read it or respond to it.

**Comment by Fish-Git** [comment]
> Update: I just downloaded the (current?) Linux source tree from https://github.com/torvalds/linux.git and the only place that RNDGETENTCNT is defined is in include/uapi/linux/random.h.
> 
> So it appears <linux/crypto.h> is the wrong header to #include. It would seem the correct header to be #included (should we even decide to "fix" it that way!) should actually be random.h (<linux/random.h>??)

FYI:

* [random(4) - Linux manual page](http://man7.org/linux/man-pages/man4/random.4.html)

So I was right: <linux/_**random.h**_> is the correct header to be #included, not <linux/_crypto.h_>.

**Comment by friedkiwi** [comment]
Ok - will update that in my copy as well.

**Comment by Fish-Git** [comment]
Fixed by commit a7d61852c5a27906eda7036b8080fe2ab7d69571.
